### PR TITLE
libs/libc:Add support for gdb rsp protocol

### PR DESCRIPTION
**Pull request by anjiahao1**
## Summary
Add support for gdb rsp protocol.
you can debug nuttx through any transport layer (serial port, network etc.),
currently supports the following functions:
1. Read and write registers
2. Read and write memory
3. Switch thread and read stack information

Future support plans:
1. Support breakpoint, watch point (requires architecture support).

related information:
https://sourceware.org/gdb/onlinedocs/gdb/Remote-Protocol.html

## Impact
debug
## Testing
with app patch [apps 1822](https://github.com/apache/nuttx-apps/pull/1822)

### CODE REVIEW

**Comment by anjiahao1** [comment]
If we can implement up_getc
Then you can cooperate with up_putc and use gdbstub to directly use gdb to locate the crash problem through the serial port

**Comment by xiaoxiang781216** [comment]
> If we can implement up_getc Then you can cooperate with up_putc and use gdbstub to directly use gdb to locate the crash problem through the serial port

Yes, please create a new patch to add up_getc/up_gets.

**Comment by xiaoxiang781216** [comment]
> `up_getc()` used to exist in the system, but was removed. I assume that this would be polled and not interrupt driven (like `up_putc()`). A polled `up_getc()` has severe consequences to system behavior, essentially locking up the system until the character is received.
> 

up_getc could call usleep to give other thread the chance to run at least in the thread context.

> Removed by commit [c83da3c](https://github.com/apache/nuttx/commit/c83da3c48fab8c4892c2d314b7ad07b8304fe574)
> 

Nice, we can revert the change to revive up_getc.
 
> An interrupt driven up_getc() that causes the caller to sleep until a character to be received would probably be usable, but not a polled version. You really can't use polled input of any kind.

for gdbstub in the kernel, the busy loop is the best choice.

**Comment by patacongo** [comment]
> > Removed by commit [c83da3c]
> 
> Nice, we can revert the change to revive up_getc.

That was back in 2016, so it will cover only the older architectures in the current source tree.  You will still have to generate many other instances of `up_getc()`.  It didn't use `up_gets()`; it used a stream, `lowinstream`, and the standard console I/O functions for input of streams with command line editing.

**Comment by patacongo** [comment]
This apps/ commit was the other part of change: 741db21f7004cf713d5d794fa2c8d72c68327e64.  It implements a version of readline() that used up_getc() in readline_common()

This could also be used with gdbstub to accomplish user-friendly command line editing.

File: include/nuttx/gdbstub/gdbstub.h [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+ * Type Definitions
+ ****************************************************************************/
+
+struct gdb_state;
+
+typedef ssize_t (*gdb_send_func)(FAR void *priv,
+                                 FAR const unsigned char *buf, size_t len);
+typedef ssize_t (*gdb_recv_func)(FAR void *priv, FAR unsigned char *buf,
+                                 size_t len);
+typedef int (*monitor_cmd_func)(FAR struct gdb_state *state, FAR char *cmd);
```
Comment (anchao) [inline include/nuttx/gdbstub/gdbstub.h:40]: *_func_t

File: include/nuttx/gdbstub/gdbstub.h [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+ *   monitor_cmd - The pointer to the monitor_cmd function.
+ *
+ * Returned Value:
+ *   The pointer to the GDB state structure on success.
+ *   NULL on failure.
+ *
+ ****************************************************************************/
+
+struct gdb_state *gdb_state_init(gdb_send_func send, gdb_recv_func recv,
+                                 void *priv, monitor_cmd_func monitor_cmd);
```
Comment (anchao) [inline include/nuttx/gdbstub/gdbstub.h:61]: move parameter priv  to last and rename monitor_cmd to monitor

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 3]
```diff
@@ -0,0 +1,1711 @@
...
+  FAR unsigned int registers[0];
+};
+
+struct gdb_handle_cmd
+{
+  CODE int (*handle)(FAR struct gdb_state *state);
+  char cmd;
+};
+
+typedef int (*gdb_enc_func)(FAR unsigned char *buf, size_t buf_len,
```
Comment (anchao) [inline libs/libc/gdbstub/lib_gdbstub.c:120]: gdb_enc_func_t

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 4]
```diff
@@ -0,0 +1,1711 @@
...
+  /* registers */
+
+  unsigned int size;
+  FAR unsigned int registers[0];
+};
+
+struct gdb_handle_cmd
+{
+  CODE int (*handle)(FAR struct gdb_state *state);
+  char cmd;
```
Comment (anchao) [inline libs/libc/gdbstub/lib_gdbstub.c:117]: char cmd;  
CODE int (*handle)(FAR struct gdb_state *state);

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 5]
```diff
@@ -0,0 +1,1711 @@
...
+ *   Zero if successful.
+ *   Negative value on error.
+ *
+ ****************************************************************************/
+
+int gdb_process(FAR struct gdb_state *state)
+{
+  int ret = 0;
+  int i;
+  static struct gdb_handle_cmd handle_cmd[] =
```
[the remark this answers is not in the document]
Comment (anjiahao1) [reply libs/libc/gdbstub/lib_gdbstub.c:1613]: In the case of some complex commands, isn't it more appropriate to use a table? , and then we implement more complex commands in different files. Later, we can also add and change tables without modifying the switch statement?

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 6]
```diff
@@ -0,0 +1,1711 @@
...
+    {
+      {
+        .handle = gdb_read_register_handle,
+        .cmd = 'g'
```
[the remark this answers is not in the document]
Comment (anjiahao1) [reply libs/libc/gdbstub/lib_gdbstub.c:1617]: i add some comment on every handle Note

**Comment by anjiahao1** [comment]
CI seems to have a problem
image
image

**Comment by pkarashchenko** [comment]
CI will be fixed by https://github.com/apache/nuttx/pull/9737

**Comment by acassis** [comment]
@anjiahao1 please take a look:
```
gdbstub/lib_gdbstub.c: In function 'get_register':
Error: gdbstub/lib_gdbstub.c:872:19: error: 'g_tcbinfo' undeclared (first use in this function)
  872 |   for (i = 0; i < g_tcbinfo.total_num; i++)
      |                   ^~~~~~~~~
gdbstub/lib_gdbstub.c:872:19: note: each undeclared identifier is reported only once for each function it appears in
gdbstub/lib_gdbstub.c: In function 'gdb_query':
Error: gdbstub/lib_gdbstub.c:1237:72: error: format '%d' expects argument of type 'int', but argument 6 has type 'size_t' {aka 'long unsigned int'} [-Werror=format=]
 1237 |       sprintf(thread_info, "Name: %s, State: %s, Priority: %d, Stack: %d",
      |                                                                       ~^
      |                                                                        |
      |                                                                        int
      |                                                                       %ld
 1238 |               tcb->name, thread_state[tcb->task_state], tcb->sched_priority,
 1239 |               tcb->adj_stack_size);
      |               ~~~~~~~~~~~~~~~~~~~                                       
      |                  |
      |                  size_t {aka long unsigned int}
gdbstub/lib_gdbstub.c: In function 'gdb_state_init':
Error: gdbstub/lib_gdbstub.c:1465:26: error: invalid use of undefined type 'const struct tcbinfo_s'
 1465 |   size_t size = g_tcbinfo.basic_num * sizeof(uintptr_t);
      |                          ^
cc1: all warnings being treated as errors
```

**Comment by xiaoxiang781216** [comment]
@anjiahao1 please normalize your config:
```
/github/workspace/sources/nuttx/tools/testbuild.sh: line 338: /github/workspace/sources/nuttx/../nuttx/nuttx.manifest: No such file or directory
  Normalize stm32f429i-disco/gdbstub
55d54
< CONFIG_SYSTEM_GDBSTUB=y
```

**Comment by anjiahao1** [comment]
> @anjiahao1 please normalize your config:
> 
> ```
> /github/workspace/sources/nuttx/tools/testbuild.sh: line 338: /github/workspace/sources/nuttx/../nuttx/nuttx.manifest: No such file or directory
>   Normalize stm32f429i-disco/gdbstub
> 55d54
> < CONFIG_SYSTEM_GDBSTUB=y
> ```

image

i use clang in local,compiled successfully
image


looks like a problem with ci

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 7]
```diff
@@ -0,0 +1,1559 @@
...
+  gdb_monitor_func_t monitor;             /* Monitor will be called when gdb
+                                           * receive a monitor command
+                                           */
+  pid_t pid;                              /* Gdb current thread */
+  FAR char *pkt_next;                     /* Pointer to next byte in packet */
+  char pkt_buf[1024];                     /* Packet buffer */
+  size_t pkt_len;                         /* Packet send and receive length */
+  uint8_t running_regs[XCPTCONTEXT_SIZE]; /* Registers of running thread */
+  size_t size;                            /* Size of registers */
+  uintptr_t registers[0];                 /* Registers of other threads */
```
Comment (pkarashchenko) [inline libs/libc/gdbstub/lib_gdbstub.c:66]: Better to use `uintptr_t registers[1];` as I'm not sure if `a[0];` is C89 compatible

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 8]
```diff
@@ -0,0 +1,1559 @@
...
+  if (buf_len * 2 < data_len)
+    {
+      return -EOVERFLOW; /* Buffer too small */
+    }
+
+  for (pos = 0; pos < data_len; pos += 2)
+    {
+      char ch[3] =
+        {
+          in[pos], in[pos + 1], 0
```
Comment (pkarashchenko) [inline libs/libc/gdbstub/lib_gdbstub.c:615]: `0` -> `\0`

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 9]
```diff
@@ -0,0 +1,1559 @@
...
+
+static int gdb_query(FAR struct gdb_state_s *state)
+{
+  static const char thread_extra_info[] = "ThreadExtraInfo";
+  static const char fthread_info[] = "fThreadInfo";
+  static const char sthread_info[] = "sThreadInfo";
+  static const char r_cmd[] = "Rcmd";
+
+  if (memcmp(&state->pkt_buf[1], fthread_info,
+              sizeof(fthread_info) - 1) == 0)
```
Comment (pkarashchenko) [inline libs/libc/gdbstub/lib_gdbstub.c:1182]: remove 1 space

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 10]
```diff
@@ -0,0 +1,1559 @@
...
+    {
+      state->pkt_buf[0] = 'm';
+      state->pkt_len = 1;
+      nxsched_foreach(gdb_get_thread, state);
+      state->pkt_len--;
+      state->pkt_buf[state->pkt_len] = 0;
+      gdb_send_packet(state);
+    }
+  else if (memcmp(&state->pkt_buf[1], sthread_info,
+                   sizeof(sthread_info) - 1) == 0)
```
Comment (pkarashchenko) [inline libs/libc/gdbstub/lib_gdbstub.c:1192]: remove 1 space

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 11]
```diff
@@ -0,0 +1,1559 @@
...
+    {
+      state->pkt_buf[0] = 'l';
+      state->pkt_len = 1;
+      gdb_send_packet(state);
+    }
+  else if (memcmp(&state->pkt_buf[1], thread_extra_info,
+                   sizeof(thread_extra_info) - 1) == 0)
```
Comment (pkarashchenko) [inline libs/libc/gdbstub/lib_gdbstub.c:1199]: remove 1 space

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 12]
```diff
@@ -0,0 +1,1559 @@
...
+static int gdb_send_stop(FAR struct gdb_state_s *state)
+{
+  return gdb_send_signal_packet(state, 0x00);
+}
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+extern const struct tcbinfo_s g_tcbinfo;
```
Comment (pkarashchenko) [inline libs/libc/gdbstub/lib_gdbstub.c:1380]: I think this should be moved to the top of the file

**Pull request merged**